doc/developer/design/20210412_postgres_sources.md

```diff
@@ -0,0 +1,210 @@
+# Postgres sources
+
+## Summary
+
+Many users are keen to use materialize with a dataset existing in a postgres
+upstream database. The current best-practice to ingest the data in realtime is
+to setup a Kafka cluster and use Debezium to convert the postgres CDC format
+into Kafka topics that can then be consumed by materialize.
+
+This design doc proposes to add a new type of source called `POSTGRES` where
+materialized directly connects to an upstream database and interprets the
+replication protocol, ingesting row updates in real time.
+
+Users should be able to try materialize out simply by running the materialize
+binary and pointing it to their postgres database, no extra infrastructure
+needed.
+
+## Goals
+
+The goal is to release a way for users to connect to a postgres database,
+covering most common setups and use cases. Since the configuration space is
+quite large this work focuses on a small set of features and makes a few
+assumptions that will get 80% of the way there.
+
+Specifically, the goals of postgres sources are to:
+
+* Connect to an upstream database with simple username/password authentication
+* Sync the initial state of the database and seamlessly switch to streaming
+* Guarantee that transaction boundaries across tables will be preserved
+* Support a suite of common column data types that cover most use cases
+* Materialize all synced tables by default. That is, all data must fit in RAM
+
+## Non-Goals
+
+In order to keep the initial scope manageable a few assumptions are made.
+Lifting those assumptions will be part of the goals for future work.
+
+* **The upstream database doesn't require complex authentication or TLS
+  connections.** After the MVP, simple TLS connections should be prioritized.
+
+* **The schema will remain static for the duration of the source.** When
+  creating a postgres source a list of upstream tables will be specified either
+  explicitly by the user or by introspecting the upstream database. After that
+  point no upstream schema changes will be supported. If such change is
+  detected, materialize will put the relevant source in an errored state and
+  wait for an administrator to fix the problem.
+
+* **All synced tables will be materialized.** Normally, sources in materialized
+  can be instantiated on demand and each instance talks to external systems
+  independently. This is a bad pattern for an upstream postgres database because
+  each instance would have to create a replication slot, which is a scarce
```

Comment:
If they're not cleaned up, replication slots can cause PG WAL files to stay around. This can result in the upstream database running out of disk space and crashing. I think it is critical to test this scenario.

Reply:
Yeah that's right. Right now we're using temporary replication slots which are tied to the current pg session and cleaned up automatically on disconnect but when we implement resumption we'll have to keep them around and clean them up on source DROP on our side.